### PR TITLE
When melding primitives, join up the type lattice

### PR DESCRIPTION
**Pull request by thatplguy**
Right now, melding an int32 primitive with another int32 primitive with a data
format will result in a conflict.  However, this should simply result in int32.

The same is true for other base types, and also between different kinds of base
types.  As another example, melding int32 and uint32 should result in int64,
not a conflict.

### CODE REVIEW

File: spec_util/meld.go
Comment (mgritter) [inline spec_util/meld.go:330]: This section looks good.

I was a bit concerned that we can't join Int64 with UInt64.  So I checked the parsing code, which for JSON numbers tries Int64 first, then Int64 again, then Uint64, then float, then bool.  (CategorizeString is called only after json.Number.Int64() fails.)

So now I'm a bit confused about how we get 32-bit integers at all?  Or don't we?  (For uploaded specs?)

It might be worth commenting above that our assumption while parsing is that everything that can fit into a signed int64 is of that type, so uint64 is really only used for values >= 2^63.

**Review by mgritter — APPROVED** [review]

File: spec_util/meld.go
Comment (liujed) [inline spec_util/meld.go:274]: This changes the advertised behaviour of `haveCompatibleTypes`. Please remember to update its documentation.

File: spec_util/meld.go
Comment (liujed) [inline spec_util/meld.go:300]: Do we want to also join ints with floating points? Maybe not if the int has a format kind attached, but it may make sense otherwise.

If we choose not to join ints with floating points, the reasoning should at least be documented.

File: spec_util/meld.go
Comment (liujed) [inline spec_util/meld.go:583]: When can we expect this to (not) return an error? When `haveCompatibleTypes(dstPrim, srcPrim)`? Comment should say. (Or better—but less scrappy—roll `haveCompatibleTypes` into this method somehow.)

File: spec_util/meld.go
Comment (liujed) [inline spec_util/meld.go:365]: The callers of `recordConflict` all set `hasConflict = true` right before the call. But if we are able to merge primitves, then there is no actual conflict. Seems like we need to either push the merging of primitives into the callers, or return a signal from `recordConflict` that no conflict actually occurred.

**Comment by liujed** [comment]
Working on a review. I think some changes need to be made. Please don't merge yet.

**Review by liujed — CHANGES_REQUESTED** [review]
Looks like we need to account for the fact that when `recordConflict` is called, a conflict no longer necessarily happens.

Additionally, I think this PR should address this TODO in `meldOneOfVariant`: https://github.com/akitasoftware/akita-libs/blob/85dbbc9c5fe1164f9b6f1265ef21adae837ce452/spec_util/meld.go#L599-L602

File: spec_util/meld.go
Comment (thatplguy) [reply spec_util/meld.go:330]: Sounds good, I added the comment.

What was your concern about not being able to join Int64 and UInt64?  That's what I had expected -- to join, we'd need a signed integer type large enough to represent all values in Int64 and UInt64, which doesn't exist in our (admittedly limited) type system.

File: spec_util/meld.go
Comment (thatplguy) [reply spec_util/meld.go:292]: Doh!  I can get rid of proto.Equal entirely.  It boils down to comparing the format kinds, but I forgot to update that part.

File: spec_util/meld.go
Comment (mgritter) [reply spec_util/meld.go:330]: I wasn't sure when we decided something was an int64 vs. an uint64.  The answer appears to be that we always assume it's signed, unless it's too large for that.  But not being able to merge -5000000000 with +5000000000 because they are a signed and unsigned int64 would have been a problem.

File: spec_util/meld.go
Comment (thatplguy) [reply spec_util/meld.go:298]: I had been thinking about data formats, but now I'm not sure there's a meaningful difference between discarding formats within the same base type vs. from in to float.  I'll extend the join.

File: spec_util/meld.go
Comment (thatplguy) [reply spec_util/meld.go:366]: Even before this change, `recordConflict` would merge primitives in some cases without introducing a `oneof`.  Looks like the existing code uses `hasConflict` to avoid merging examples on conflicts, but `recordConflict` takes care of merging examples when there isn't a conflict.

I'm not going to refactor that as part of this PR.

File: spec_util/meld.go
Comment (thatplguy) [reply spec_util/meld.go:584]: I considered it but didn't want to do more refactoring.  `meldPrimitive` is called exactly once and isn't exposed outside this package, so I'm not too worried.  I updated the comment.

**Review by liujed — COMMENTED** [review]
This TODO isn't addressed yet: https://github.com/akitasoftware/akita-libs/blob/85dbbc9c5fe1164f9b6f1265ef21adae837ce452/spec_util/meld.go#L599-L602

(GitHub won't let me point to it like in the other comments because the PR doesn't touch that part of the code.)

**Comment by thatplguy** [comment]
Thanks, forgot about that.  TODO is fixed in 2e2c30c.

File: spec_util/meld_test.go
Comment (liujed) [inline spec_util/meld_test.go:588]: I've seen meld tests, like the one below, that suggest that meld should be commutative. Why not test both directions here instead of having two copies of the test?

https://github.com/akitasoftware/akita-libs/blob/85dbbc9c5fe1164f9b6f1265ef21adae837ce452/spec_util/meld_test.go#L239-L270

File: spec_util/meld_test.go
Comment (liujed) [inline spec_util/meld_test.go:294]: Similarly for this test.

**Review by liujed — APPROVED** [review]
Small nit regarding tests, but looks good.

File: spec_util/meld_test.go
Comment (thatplguy) [reply spec_util/meld_test.go]: Good idea, fixed in 3089a2d.

**Pull request merged**
